xray/conventions.py

```diff
@@ -53,11 +55,18 @@ def mask_and_scale(array, fill_value=None, scale_factor=None, add_offset=None,
     """
     # by default, cast to float to ensure NaN is meaningful
     values = np.array(array, dtype=dtype, copy=True)
-    if fill_value is not None and not pd.isnull(fill_value):
-        if values.ndim > 0:
-            values[values == fill_value] = np.nan
-        elif values == fill_value:
-            values = np.array(np.nan)
+    if fill_value is not None and not np.all(pd.isnull(fill_value)):
+        if isinstance(fill_value, np.ndarray) and len(fill_value) != 1:
+            # multiple values in fill_value
+            warnings.warn('fill_value contains multiple values, decoding all '
```

Comment:
How about moving this warning to inside `decode_cf_variable`? That would also let us say something more explicit like: `"variable 'temperature' has multiple fill values [0, 10], decoding all values to NaN`

That way, the warning appears only once (per variable) when opening a file, not repeatedly when reading the values.